src/preamble.js

```diff
@@ -802,14 +801,81 @@ var wasmOffsetConverter;
 #if SPLIT_MODULE
 var splitModuleProxyHandler = {
   'get': function(target, prop, receiver) {
-    err('placeholder function created: ' + prop);
     return function() {
-      abort('placeholder function called: ' + prop);
+      err('placeholder function called: ' + prop);
+      var imports = {'primary': Module['asm']};
+      var table = Module['asm']['__indirect_function_table'];
+      instantiateSync(wasmBinaryFile + '.deferred', imports);
+      err('instantiated deferred module, continuing');
+      return table.get(prop).apply(null, arguments);
     }
   }
 };
 #endif
 
+#if SPLIT_MODULE || !WASM_ASYNC_COMPILATION
+function instantiateSync(file, info) {
+  var instance;
+  var module;
+  var binary;
+  try {
+    binary = getBinary(file);
+#if NODE_CODE_CACHING
+    if (ENVIRONMENT_IS_NODE) {
+      var v8 = require('v8');
+      // Include the V8 version in the cache name, so that we don't try to
+      // load cached code from another version, which fails silently (it seems
+      // to load ok, but we do actually recompile the binary every time).
+      var cachedCodeFile = file + v8.cachedDataVersionTag() + '.cached';
+      cachedCodeFile = locateFile(cachedCodeFile);
+      if (!nodeFS) nodeFS = require('fs');
+      var hasCached = nodeFS.existsSync(cachedCodeFile);
+      if (hasCached) {
+#if RUNTIME_LOGGING
+        err('NODE_CODE_CACHING: loading module');
+#endif
+        try {
+          module = v8.deserialize(nodeFS.readFileSync(cachedCodeFile));
+        } catch (e) {
+          err('NODE_CODE_CACHING: failed to deserialize, bad cache file? (' + cachedCodeFile + ')');
+          // Save the new compiled code when we have it.
+          hasCached = false;
+        }
+      }
+    }
+    if (!module) {
+      module = new WebAssembly.Module(binary);
+    }
+    if (ENVIRONMENT_IS_NODE && !hasCached) {
+#if RUNTIME_LOGGING
+      err('NODE_CODE_CACHING: saving module');
+#endif
+      nodeFS.writeFileSync(cachedCodeFile, v8.serialize(module));
+    }
+#else // NODE_CODE_CACHING
+    module = new WebAssembly.Module(binary);
+#endif // NODE_CODE_CACHING
+    instance = new WebAssembly.Instance(module, info);
+#if USE_OFFSET_CONVERTER
+    wasmOffsetConverter = new WasmOffsetConverter(binary, module);
+    {{{ runOnMainThread("removeRunDependency('offset-converter');") }}}
+#endif
+  } catch (e) {
+    var str = e.toString();
+    err('failed to compile wasm module: ' + str);
+    if (str.indexOf('imported Memory') >= 0 ||
+        str.indexOf('memory import') >= 0) {
+      err('Memory size incompatibility issues may be due to changing INITIAL_MEMORY at runtime to something too large. Use ALLOW_MEMORY_GROWTH to allow any size memory (and also make sure not to set INITIAL_MEMORY at runtime to something smaller than it was at compile time).');
+    }
+    throw e;
+  }
+#if LOAD_SOURCE_MAP
+  receiveSourceMapJSON(getSourceMap());
+#endif
+  return {'instance': instance, 'module': module};
```

Comment:
Is there a more code size efficient way to return a pair?   Also is it going to confuse closure that you use string keys here but non-string keys when you receive this object?

Reply:
I'm not sure on both counts! Maybe @kripken knows?

Reply:
You could do `[instance, module]` but it would be an array, and lack the names, which I think we need here?

Reply:
Why do you need the names?   Returning an array pair seems more compact.